services/domainService.ts

```diff
@@ -0,0 +1,177 @@
+import { Context, HttpRequest } from "@azure/functions";
+import {
+  Address,
+  DomainFindOptions,
+  DomainId,
+  Maybe,
+} from "@zero-tech/data-store";
+import { MongoDbService } from "@zero-tech/data-store/lib/database/mongo/mongoDbService";
+import { MongoClient } from "mongodb";
+import { getConnectedDbClient, getDatabaseService } from "../helpers";
+import { validateBlockchainAddress, validateDomainId } from "../schemas";
+
+// add try/catch/finally wrapper for 500s
+export async function doOp(
+  serviceFn: Function,
+  context: Context,
+  request: HttpRequest
+) {
+  let dbClient: Maybe<MongoClient>;
+
+  try {
+    dbClient = await getConnectedDbClient(context);
+    const dbService: MongoDbService = getDatabaseService(dbClient);
+    await serviceFn(context, request, dbService);
+  } catch (e: any) {
+    const error: Error = {
+      name: e.name ?? "Unknown Internal Server Error",
+      message: e.message ?? "An unexpected server error occured.",
+    };
+    context.log(e.name, e.message, e.stack);
+
+    context.res = {
+      status: 500,
+      body: `${error.name} \n ${error.message}`,
+    };
+  } finally {
+    if (dbClient) {
+      await dbClient.close();
+    }
+  }
+}
+
+export async function listDomains(
+  context: Context,
```

Comment:
Not that I think it needs to be fixed in this PR. But this is introducing Azure Functions code smell into `services/`

ie: `Context` and `HttpRequest` are specific for azure functions.

Reply:
Related to my comment:
I think Context and HttpRequest should be kept at the top level of the azure function. All the Httprequest handling code should be left there, so the services can be lean and more cohesive

Reply:
@Will-Hyde:
Perhaps a better name for `domainService` would have been `domainHttpService`. It's an abstraction that performs a domain HTTP operation after being triggered by an HTTP trigger function. The actual interaction with the data models is handled by the abstraction provided by the `data-store` pkg.

Could you help me understand the benefits? It feels like things become less lean when we move request/response handling up to the httpTrigger level. Every azure function now needs to handle getting a connection, have it's own try/catch internal server error handling and close its own database connection.

@Remscar Could you elaborate on why these are code smells? I took this pattern from a Microsoft [guide](https://docs.microsoft.com/en-us/learn/modules/shift-nodejs-express-apis-serverless/6-exercise-migrate-services-azure-functions) for migrating APIs to Azure. The context object also includes a `logger`, which seemed useful for creating traces down the stack.

Reply:
I think this is fine for now. Would be better to specify this is adapter code specifically for Azure Functions. But all business logic is abstracted and code-smell free in the package this code uses.

Essentially code smell is when you use external code and then build your code around it. Instead of building adapaters/helpers to handle those external types, and then you translate to/from those types when needed.